include/dmlc/logging.h

```diff
@@ -428,30 +426,42 @@ class LogMessageFatal : public LogMessage {
 #else
 class LogMessageFatal {
  public:
-  LogMessageFatal(const char* file, int line) {
-    log_stream_ << "[" << pretty_date_.HumanDate() << "] " << file << ":"
-                << line << ": ";
+  LogMessageFatal(const char *file, int line) {
+    Entry::ThreadLocal()->Init(file, line);
   }
-  std::ostringstream &stream() { return log_stream_; }
-  ~LogMessageFatal() DMLC_THROW_EXCEPTION {
+  std::ostringstream &stream() { return Entry::ThreadLocal()->log_stream; }
+  DMLC_NO_INLINE ~LogMessageFatal() DMLC_THROW_EXCEPTION {
 #if DMLC_LOG_STACK_TRACE
-    log_stream_ << "\n" << StackTrace(1, LogStackTraceLevel()) << "\n";
-#endif
-
-    // throwing out of destructor is evil
-    // hopefully we can do it here
-    // also log the message before throw
-#if DMLC_LOG_BEFORE_THROW
-    LOG(ERROR) << log_stream_.str();
+    Entry::ThreadLocal()->log_stream << "\n"
+                                     << StackTrace(1, LogStackTraceLevel())
+                                     << "\n";
 #endif
-    throw Error(log_stream_.str());
+    throw Entry::ThreadLocal()->Finalize();
   }
 
  private:
-  std::ostringstream log_stream_;
-  DateLogger pretty_date_;
-  LogMessageFatal(const LogMessageFatal&);
-  void operator=(const LogMessageFatal&);
+  struct Entry {
+    std::ostringstream log_stream;
+    DMLC_NO_INLINE void Init(const char *file, int line) {
+      DateLogger date;
+      log_stream.str("");
+      log_stream.clear();
+      log_stream << "[" << date.HumanDate() << "] " << file << ":" << line
+                 << ": ";
+    }
+    dmlc::Error Finalize() {
+#if DMLC_LOG_BEFORE_THROW
+      LOG(ERROR) << log_stream.str();
+#endif
+      return dmlc::Error(log_stream.str());
+    }
+    DMLC_NO_INLINE static Entry *ThreadLocal() {
+      static thread_local Entry *result = new Entry();
```

Comment:
The introduction of TLS helps to reduce the size of class `dmlc::LogMessageFatal` to zero, and thus enables some opportunities in compiler optimization